chore: create upgrade guide (#2494)
hashicorp/terraform-cdk: website/docs/cdktf/release/upgrade-guide-v0-15.mdx

---
page_title: Upgrading to CDKTF Version 0.15
description: >-

>   Minimum version of Terraform increased to 1.2, classes reanamed for C#, Java, and Go

Reviewer: ```suggestion
  Minimum version of Terraform increased to 1.2, some interfaces renamed for C#, Java, and Go
```

---

# Upgrading to CDK for Terraform Version 0.15

0.15 is primarily focused around CLI improvements targeted at a better Terraform Cloud integration.

> Because of this, there are only smaller breaking changes in this release. The biggest one is a refactoring of several Constructs in our core library.

Reviewer: I'd make it a bit more weak, as this is a minor breaking change but the sentence sounded a bit extreme 😁
Maybe something like this?
```suggestion
Because of this, there are only smaller breaking changes in this release. The most notable one is a renaming of some interfaces in our core library.
```

Some configuration classes were suffixed inconsistently with `Config`, `Props` or `Options` and were renamed to `Config` for consistency. These classes need to be used in C#, Java, and Go, therefore we have a breaking change in these languages.

## Minimum version of Terraform increased to 1.2

CDKTF is bumping the minimum supported version of Terraform from 1.0 to 1.2 starting from CDKTF version 0.15. This change is necessary as CDKTF 0.15 uses the Terraform CLI for planning and applying changes for Terraform Cloud and Terraform Enterprise instead of [the API](https://developer.hashicorp.com/terraform/cloud-docs/api-docs) used in version 0.14.

[Terraform promises compatibility](https://developer.hashicorp.com/terraform/language/v1-compatibility-promises) between all v1.x releases, so there shouldn't be any issues with this change. The only changes between these versions are bug fixes, and new feature additions.

Please open an issue if you encounter any issues or bugs with this change.

## Renaming of `*Options` and `*Props` to `*Config`

> We had several exports where the configuration passed to constructs or functions was using a class with a different suffix then `Config`. If you are using a language like C#, Java, Python, or Go you might have needed to specify these classes in your CDKTF application. To simplify the usage we changed every suffix to be `Config`, so if you currently use one of the others please rename them to use the `Config` suffix. No options were changed in the process, only the names were aligned.

Reviewer: this comment is naming Python (and we have no example) while all other mentions of languages above don't include Python – were there changes to Python? I suspect there were some, as you sometimes have to use some interface names for Python?

Author: Removing Python, I don't think one needs to specify them in python

### C#

```diff
using vpc; // a local module
class MyStack : TerraformStack
  {
      public MyStack(Construct scope, string name) : base(scope, name)
      {
-          Vpc vpc = new Vpc(this, "vpc", new VpcOptions {

> +          Vpc vpc = new Vpc(this, "vpc", new VpcConf {

Reviewer: ```suggestion
+          Vpc vpc = new Vpc(this, "vpc", new VpcConfig {
```

              Name = "vpc",
              PublicSubnets = new string[] {"10.0.1.0/24", "10.0.2.0/24"}
          });
      }
  }

```

### Java

```diff
import software.constructs.Construct;
import com.hashicorp.cdktf.App;
import com.hashicorp.cdktf.TerraformStack;
import com.hashicorp.cdktf.CloudBackend;
-import com.hashicorp.cdktf.CloudBackendProps;
+import com.hashicorp.cdktf.CloudBackendConfig;
import com.hashicorp.cdktf.NamedCloudWorkspace;

    public static class Demo extends TerraformStack{

        public Demo(Construct scope, String id){
            super(scope, id);
-            new CloudBackend(this, CloudBackendProps.builder()
+            new CloudBackend(this, CloudBackendConfig.builder()
                    .organization("hashicorp")
                    .workspaces(new NamedCloudWorkspace("demo"))
                    .build()
            );
```

> ### Go

Reviewer: ehhh, below is Java code 😅

```diff
import com.hashicorp.cdktf.App;
import com.hashicorp.cdktf.TerraformHclModule;

> import com.hashicorp.cdktf.TerraformHclModuleOptions;

Reviewer: this one needs a diff line, too. `Options` is old.
